### PR TITLE
Join existing list

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person
## Summary
We want to allow users to join existing shopping lists so they can manage them with friends or family. This would take the following steps:
1. User enters an existing token into a form
2. The token is saved to `localStorage`
3. The user is taken to the `List` view and shown the existing list.

## Acceptance criteria
If a user **doesn’t** already have a token:
- [ ] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [ ] The input that accepts the list token has a semantic `label` element associated with its
- [ ] The user can submit this form with both the mouse and the `Enter` key
- [ ] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [ ] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [ ] They are automatically redirected to the `List` view.

## Notes
- Be sure to use the `setListToken` function to save the token that the user enters!
- You will know your `<label>` is correctly implemented if you click on the label and keyboard focus moves to the related input. Refer to [the MDN docs on `<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) for more info.
- Users will be able to submit the form with the `Enter` key if you listen for the right JavaScript event on the right element! If you're adding an `onKeyDown` listener anywhere, you might be overcomplicating things.
- You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.
- If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 3).

### PR DESCRIPTION
**Pull request by tinazarb**
## Description

Allows users to join an existing shopping list

## Related Issue

Closes #5 

## Acceptance Criteria

If a user **doesn’t** already have a token:
- [X] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [X] The input that accepts the list token has a semantic `label` element associated with its
- [X] The user can submit this form with both the mouse and the `Enter` key
- [X] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [X] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [X] They are automatically redirected to the `List` view.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
Screen Shot 2023-01-26 at 3 55 59 PM

### After
Screen Shot 2023-01-26 at 3 57 00 PM

## Testing Steps / QA Criteria
Type an existing list token into the Enter Token form. Verify the token re-directs to the existing list and appears in local storage.

### CODE REVIEW

File: src/views/Home.jsx [hunk 1]
```diff
@@ -1,12 +1,58 @@
...
 	return (
 		<div className="Home">
 			<p>
 				Hello from the home (<code>/</code>) page.
 			</p>
-			<button onClick={handleNewToken}>Create a new list!</button>
+			<button onClick={handleClick}>Create a new list!</button>
+			<p> Want to join an existing list? </p>
+			<form onSubmit={handleTokenSubmit}>
+				<label htmlFor="token"> Enter Token:</label>
```
Comment (DwightTheShark) [inline src/views/Home.jsx:45]: I know it's a token, but is this accessible for someone who isn't computer literate?  Just a thought, you probably already had it too.

**Review by DwightTheShark — APPROVED** [review]

File: src/api/firebase.js [hunk 2]
```diff
@@ -79,3 +85,10 @@ export async function deleteItem() {
 	 * this function must accept!
 	 */
 }
+
+export async function checkListToken(listId) {
```
Comment (ehammes) [inline src/api/firebase.js:89]: @tinazarb @louieknolle Did you determine what was needed here via the Firebase docs?

File: src/api/firebase.js [hunk 3]
```diff
@@ -79,3 +85,10 @@ export async function deleteItem() {
...
+	const listTokenQuery = query(collection(db, listId));
+	const listTokenSnapshot = await getDocs(listTokenQuery);
+
```
Comment (ehammes) [inline src/api/firebase.js:92]: @tinazarb @louieknolle Did you determine what was needed here via the Firebase docs?

File: src/views/Home.jsx [hunk 4]
```diff
@@ -1,12 +1,58 @@
...
+				<input
+					type="text"
+					name="token"
+					id="token"
+					value={token}
```
Comment (ehammes) [inline src/views/Home.jsx:50]: Edge case...do we need to consider case sensitivity? ie token is 'my test list'....user enters 'My Test List'
or is the assumption that the user knows that it must match exactly, similar to a password. Could use the toLowerCase() method if its something we decide to consider.

Comment (tinazarb) [reply src/views/Home.jsx:50]: @ehammes theoretically both "my test list" and "My test list" could be two different lists in the database and both exist

Comment (ehammes) [reply src/views/Home.jsx:50]: Future improvement...may want to account for form validation here if the user clicks 'Join' without an input value as well as catch all error (ie Firebase errors, etc.)

File: src/api/firebase.js (hunk 3 above)
Comment (tinazarb) [reply src/api/firebase.js:92]: @ehammes yes we are querying the database to see if it the collection has documents associated with it as that's required, and it returns false if there aren't/collection doesn't exist

File: src/views/Home.jsx [hunk 5]
```diff
@@ -1,12 +1,58 @@
...
+import { useNavigate } from 'react-router-dom';
+
+export function Home({ handleNewToken, setListToken, listToken }) {
+	const [token, setToken] = useState('');
+	const [listNotFound, setlistNotFound] = useState('');
+	const redirect = useNavigate();
+
+	if (listToken) {
+		redirect('/list');
+	}
```
Comment (ehammes) [inline src/views/Home.jsx:13]: @tinazarb @louieknolle Future improvement we may want to consider... currently if I join an existing list then I can't switch to another existing list unless I clear out local storage manually.

**Review by ehammes — APPROVED** [review]
Nice work! Just a few comments with questions/future improvements to consider. LGTM!

File: src/views/Home.jsx (hunk 1 above)
Comment (tinazarb) [reply src/views/Home.jsx:45]: Good point! Updated with aria-describedby description as per @mxmason

File: src/views/Home.jsx (hunk 4 above)
Comment (timothy-taylor) [reply src/views/Home.jsx:50]: Always a great idea to validate your input. Checkout the built-in [validations](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#using_built-in_form_validation) as a first line of defense. You can then read up on using javascript for input validation, but at the very least normalizing your strings in someway will probably be a good idea.

File: src/views/Home.jsx [hunk 6]
```diff
@@ -1,12 +1,63 @@
...
+					type="text"
+					name="token"
+					id="token"
+					value={token}
+					onChange={handleFormChange}
+					aria-describedby="token-desc"
+				/>
+				<button type="submit"> Join</button>
+				<div id="token-desc">
+					A token is three space-separated words, like{' '}
```
Comment (timothy-taylor) [inline src/views/Home.jsx:57]: Love this!

**Review by timothy-taylor — APPROVED** [review]
Great job!

File: src/views/Home.jsx (hunk 4 above)
Comment (tinazarb) [reply src/views/Home.jsx:50]: Copy! Added in a required form validation and accounted for token case insensitivity

**Review by timothy-taylor — APPROVED** [review]
